src/js/plugins/tts/AsyncStream.js

```diff
@@ -0,0 +1,212 @@
+/**
+ * @export
+ * The underlying object wrapped each stream's elements. (Based on
+ * JavaScript Generators; see
+ * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator/next )
+ * @template T
+ * @typedef {Object} TStreamItem
+ * @property {T} value
+ * @property {Boolean} done
+ */
+
+/**
+ * @abstract
+ * Class that makes it easier to deal with a series of async events
+ * (represented as Promises). All the function let you interact with
+ * the promise's value instead of the promise itself, letting you 
+ * treat it as a stream of synchronuous values. The notable exception
+ * being the `pull` method, which should be used to consume the stream,
+ * and hence must return the promise directly.
+ * @template T
+ */
+export default class AsyncStream {
+    /** @typedef {TStreamItem<T>} TItem */
```

Comment:
This looks like it'd be a good use case for TypeScript generics 😉  _hides_ 🙈

Reply:
Hahaha; vs code does autocompletion for these, too! I broke it with the `flatten` method, but everything else here typechecks. If you add `//@ts-check` to the top of a JS file, it'll **highlight all the type violations** 😍